[easy] [typetraits] add `==` for type equality (different from `is` !)

timotheecour opened the pull request.

hlaaftana commented on lib/pure/typetraits.nim:
```diff
@@ -57,6 +57,16 @@ proc supportsCopyMem*(t: typedesc): bool {.magic: "TypeTrait".}
   ## This trait returns true iff the type ``t`` is safe to use for
   ## `copyMem`:idx:. Other languages name a type like these `blob`:idx:.
 
+proc sameType*(t1, t2: typedesc): bool =
```
I predict the same thing with `name` will happen where people will ask for this to be renamed to `==`

timotheecour replied:
i thought about it... not sure which is better; sameType is more searchable but `==` is more predictable; Note: there's also `macros.sameType(Node,Node)` although obviously that one can't be named to `==`.
Overall `==` could be better ; should i go ahead and chage to `==`?

timotheecour replied:
done. replace with `==` ; which is nice because we get `int != float` for free.
makes the language nice and uniform

Araq replied:
This will quickly turn into ``compareTypes(a, b, flags)`` anyway, ``==`` is for simple comparisons, type equality is anything but simple. In fact, I know few other examples where it gets as complex as it gets for types.

timotheecour replied:
well nothing wrong w having both `==` for simple comparison and `compareTypes(a, b, flags)` to allow customizing what is meant by equality

timotheecour commented:
PTAL

dom96 commented:
How is this different to ``is``? 

This passes:

```nim
type T = int
doAssert T is int
doAssert int is T
doAssert int isnot float
```

GULPF commented:
@dom96 `is` doesn't imply equality since it supports concepts & inheritance.

```nim
type
  C = concept x
    x is int

echo int is C # true
echo C is int # false
```

timotheecour commented:
clarified documentation further to add what @GULPF said.
PTAL

dom96 commented:
This should be demonstrated in the runnable examples.

dom96 commented on lib/pure/typetraits.nim:
```diff
@@ -57,6 +57,16 @@ proc supportsCopyMem*(t: typedesc): bool {.magic: "TypeTrait".}
+  ## Returns wheter ``t1``, ``t2`` are the same type; this is different
+  ## from ``t1 is t2`` since the latter supports concepts & inheritance.
+  runnableExamples:
+    type T=int
```
Style nitpick: space before and after ``=``.

timotheecour replied:
done

dom96 commented on lib/pure/typetraits.nim:
```diff
@@ -57,6 +57,16 @@ proc supportsCopyMem*(t: typedesc): bool {.magic: "TypeTrait".}
+    type T=int
+    doAssert T == int
+    doAssert int == T
+    doAssert: int != float
```
Please be consistent in your style, no need for this random ``:``.

timotheecour replied:
it's needed to avoid https://github.com/nim-lang/Nim/issues/8300
otherwise you get: `Error: invalid indentation`

Araq replied:
I doubt it. ``!=`` is not an unary keyword operator.

timotheecour replied:
you can try for yourself with `git fetch origin pull/8634/head:pull_8634 && git checkout pull_8634` ...

Araq replied:
This compiles:

```nim

import typetraits

type T = int

template `==`(x, y: typedesc): bool = (x is y) and (y is x)

doAssert T == int
doAssert int == T
doAssert int != float
```

timotheecour replied:
ok, this must've been fixed since that was reported

timotheecour commented:
PTAL, all comments addressed

timotheecour commented:
/cc @dom96 ping

Araq commented:
I don't want it. People should use the ``is`` operator and learn Nim. There are two ways to design a programming language:

1. Keep it simple, well documented and expect people to read the manual.
2. Borrow everything that might be useful from other languages so that programmers can guess things correctly.

(2) is a never ending story and in the end you fail to produce a consistent system that people enjoy using after they got over the "I program in Nim for X days" phase. It's also after this phase when programmers become valuable contributors, generally speaking. We tried (2) for far too long, it's time to try (1).

timotheecour commented on lib/pure/typetraits.nim:
```diff
@@ -57,6 +57,16 @@ proc supportsCopyMem*(t: typedesc): bool {.magic: "TypeTrait".}
   ## `copyMem`:idx:. Other languages name a type like these `blob`:idx:.
 
+proc `==`*(t1, t2: typedesc): bool =
+  ## Returns wheter ``t1``, ``t2`` are the same type; this is different
```
done

timotheecour commented on lib/pure/typetraits.nim:
```diff
@@ -57,6 +57,16 @@ proc supportsCopyMem*(t: typedesc): bool {.magic: "TypeTrait".}
 
+proc `==`*(t1, t2: typedesc): bool =
+  ## Returns wheter ``t1``, ``t2`` are the same type; this is different
+  ## from ``t1 is t2`` since the latter supports concepts & inheritance.
```
it does:
```nim
type A=object of RootObj
  b1:int
type B=object of A
  b2:int

echo A is B # false
echo B is A # true
```

bluenote10 commented:
In my opinion having this function is a nice way to document this subtlety. Seeing the expression `A is B and B is A` in the wild would often raise the question why this is necessary. It wasn't even obvious to @dom96 and currently the manual section on the is operator also does not mention it. Implementing `==` allows to document this difference nicely in code + docstring. What's the drawback?

Araq commented:
> What's the drawback?

It doesn't come up often enough. Most of the time people would use it instead of ``is`` because they are not aware of ``is``.

timotheecour commented:
* I changed implementation to use `macros.sameType` because some test cases were running into https://github.com/nim-lang/Nim/issues/8693
this introduces a dependency `typetraits: import macros`; if that's a problem I could move `==` to `macros.nim` instead although it seems more natural in typetraits.nim 

* I added a bunch of tests (see tests/metatype/ttypetraits2.nim)

* I don't really see how https://github.com/nim-lang/Nim/pull/8612#issuecomment-414078906 applies to this PR. `is` is different from `==`, there are use cases for both.

> It doesn't come up often enough. Most of the time people would use it instead of is because they are not aware of is.

I'm using both `is` and this PR's `==` quite a bit in my own code when working, in particular in tests. I don't follow this argument; `is` and `==` are 2 different tools with 2 different use cases; we shouldn't assume nim programmers are incapable of telling these apart ; so long documentation makes it clear, there is no drawback and it makes the language more regular / self-consistent. This has nothing to do with "Borrow everything that might be useful from other languages"

Clyybber commented:
https://github.com/nim-lang/Nim/pull/8700 would fix #8693, so I think this PR should use the simpler old implementation.

timotheecour commented:
happy to update if/when #8700 gets merged (looks like tests in https://github.com/nim-lang/Nim/pull/8700 are failing right now, and there are unresolved comments there)

Clyybber commented:
The tests in #8700 are passing now.

timotheecour commented:
> #8700 would fix #8693, so I think this PR should use the simpler old implementation.


done

Araq commented:
Well I was asking for concrete examples and you provided one where I have to ask: What's the point of the check? ``when isValueType(T1)`` seems to suffice.

Araq commented:
Your check for ``T1 == T2`` doesn't fix your wrong ``isValueType`` predicate. I'm asking for real examples, I'm not asking for your workarounds or why ``==`` is different than ``is`` (I know!).

And the reason why I don't like this feature is because this whole idea of builtin specialized typetraits does not scale. What should be done instead is to make use of ``macros.getType()`` so that one can write real checks for the type properties one is interested in. Your ``isValueType`` is a good example here, you don't want ``isValueType``, you want ``isBlob`` or similar.

timotheecour commented:
> you don't want isValueType, you want isBlob or similar.

I couldn't find isBlob in nim; so perhaps we actually mean the same thing? see below.

> wrong isValueType predicate

without seeing implementation it's hard to say it's wrong, so here is my implementation for `isValueType`: (the name could be better; it basically means it's a POD (plain old data); ie doesn't any pointers / references)

given this implementation, why is it wrong? (besides the TODO of using a whitelist instead of `not(tuple|object|array|RefType)` and fixing remaining cases of map etc)

i think it's a real example; it tries to handle as much cases as possible with bitEquals, reverting to manual behavior otherwise. it's not a workaround (of what?)

```nim
{.reorder:on.}
proc defaultTemp[T]():T=
  var a: T
  return a

# TODO: map? etc
type RefType = ptr|pointer|ref|seq|string
proc isValueType(T:type RefType): bool = false

proc isValueType(T:type[tuple|object]): bool =
  ## returns whether T is a POD type
  var a = defaultTemp[T]()
  for i in fields(a):
    if not isValueType(type(i)):
      return false
  return true

# proc isValueType(T:type array):bool = defaultTemp[T]()[0].type.isValueType
#or: 
proc isValueType[N, T2](T:type[array[N, T2]]):bool = isValueType(T2)

# TODO: use whitelist instead (more robust)
proc isValueType(T:type[not(tuple|object|array|RefType)]):bool = true


proc isValueType2*(T:type): bool = isValueType(T)

proc isValueTypeTest()=
  type A1=object
    a:int
    b:array[3, int]
    c:type((1,2))

  type A2=object
    a:int
    b:array[3, string]

  var a:A1
  var a2 = cast[pointer](a.addr)

  require:
    int.isValueType
    float.isValueType
    not string.isValueType
    not seq[int].isValueType
    (1,2).type.isValueType
    not (1,"").type.isValueType
    A1.isValueType
    not A2.isValueType
    not (ref A1).isValueType
    not (ptr int).isValueType
    a.type.isValueType
    not a.addr.type.isValueType
    not a2.type.isValueType
    not pointer.isValueType
  # make sure works at CT
  static:
    doAssert: not A2.isValueType
```

Araq commented:
Still not convinced.
